pyomo/contrib/alternative_solutions/solnpool.py

```diff
@@ -9,119 +9,1004 @@
+        return args[0]
+    return Solution(*args, **kwargs)
 
-def gurobi_generate_solutions(
-    model,
-    *,
-    num_solutions=10,
-    rel_opt_gap=None,
-    abs_opt_gap=None,
-    solver_options={},
-    tee=False,
-):
+
+def _as_pyomo_solution(*args, **kwargs):
+    if len(args) == 1 and len(kwargs) == 0:
+        assert type(args[0]) is Solution, "Expected a single solution"
+        return args[0]
+    return PyomoSolution(*args, **kwargs)
+
+
+class PoolCounter:
+    """
+    A class to wrap the counter element for solution pools.
+    It contains just the solution_counter element.
+    """
+
+    solution_counter = 0
+
+
+class SolutionPoolBase:
     """
-    Finds alternative optimal solutions for discrete variables using Gurobi's
-    built-in Solution Pool capability.
+    A class to manage groups of solutions as a pool.
+    This is the general base pool class.
 
-    This method defaults to the optimality-enforced discovery method with PoolSearchMode = 2.
-    There are two other options, standard single optimal solution (PoolSearchMode = 0) and
-    best-effort discovery with no guarantees (PoolSearchMode = 1). Please consult the Gurobi
-    documentation on PoolSearchMode for details on impact on Gurobi results.
-    Changes to this mode can be made by included PoolSearchMode set to the intended value
-    in solver_options.
+    This class is designed to integrate with the alternative_solution generation methods.
+    Additionally, groups of SolutionPool objects can be handled with the PoolManager class.
 
     Parameters
     ----------
-    model : ConcreteModel
-        A concrete Pyomo model.
-    num_solutions : int
-        The maximum number of solutions to generate. This parameter maps to
-        the PoolSolutions parameter in Gurobi.
-    rel_opt_gap : non-negative float or None
-        The relative optimality gap for allowable alternative solutions.
-        None implies that there is no limit on the relative optimality gap
-        (i.e. that any feasible solution can be considered by Gurobi).
-        This parameter maps to the PoolGap parameter in Gurobi.
-    abs_opt_gap : non-negative float or None
-        The absolute optimality gap for allowable alternative solutions.
-        None implies that there is no limit on the absolute optimality gap
-        (i.e. that any feasible solution can be considered by Gurobi).
-        This parameter maps to the PoolGapAbs parameter in Gurobi.
-    solver_options : dict
-        Solver option-value pairs to be passed to the Gurobi solver.
-    tee : boolean
-        Boolean indicating that the solver output should be displayed.
-
-    Returns
-    -------
-    solutions
-        A list of Solution objects.  [Solution]
+    name : String
+        String name to describe the pool.
+    as_solution : Function or None
+        Method for converting inputs into Solution objects.
+        A value of None will result in the default _as_solution function being used.
```

Comment:
I like this callback design for its extensibility (and because in a lot of contexts we really only care about part of a solution, not every Var in the model, and this will allow for defining that as a "solution" for AOS purposes.). Can you document somewhere what the expected input and output for this function are? It seems like it is assumed it returns a `Solution` object, right? And it can actually be called given anything because it will pass through what is given to `add`?